### PR TITLE
1.29.0-0.4.1: Allow Hardware wallets to add custom network info

### PR DESCRIPTION
**Pull request by matextrem**
As a consequence of this [issue](https://github.com/gnosis/safe-react/issues/1532) we need a way to get Hardware wallets to work with NON-ethereum networks. 

Our proposal is the following: 

- When a user instantiates Onboard in a dApp and add a hardware wallet

```
const onboard = Onboard({
   networkId: 1234
  //... other options
  walletSelect: {
     wallets:[
         {
             walletName: 'ledger',
             rpcUrl: RPC_URL,
            // We add a new custom parameter in order to specify the info needed for a custom network to work with HW 
            wallets.
             customNetwork: {
                  networkId: number
                  genesis: GenesisBlock
                  hardforks: Hardfork[]
                  bootstrapNodes: BootstrapNode[]
            }
         },
     ] 
  }
})
```
So basically, if customNetwork parameter is specified, it will be used for @ethereumjs-tx lib to create and sign the TX. If not, it will only used the networkId specified at the top level of the root object.

### CODE REVIEW

File: src/modules/check/derivation-path.ts [hunk 1]
```diff
@@ -224,6 +224,7 @@ function derivationPath(
                   state.completed = true
                 })
                 .catch(error => {
+                  console.log(error)
```
Comment (nicosampler) [inline src/modules/check/derivation-path.ts:227]: I think this should be removed.

File: src/modules/select/wallets/hd-wallet.ts [hunk 2]
```diff
@@ -46,16 +45,16 @@ export function isValidPath(path: string) {
     return false
   }
 
-  if (parts[2] !== "60'" && parts[2] !== "1'") {
+  if (parts[2] !== "60'" && parts[2] !== "1'" && parts[2] !== "73799'" && parts[2] !== "246'") {
```
Comment (nicosampler) [inline src/modules/select/wallets/hd-wallet.ts:48]: perhaps
```
if(!["60'", "1'", "73799'",  "246'"].includes(parts[2])) {
 ...
}
```

File: src/modules/select/wallets/ledger.ts [hunk 3]
```diff
@@ -205,6 +208,7 @@ async function ledgerProvider(options: {
         ? LedgerTransport.listen(observer)
         : TransportU2F.listen(observer)
     } catch (error) {
+      console.log(error)
```
Comment (nicosampler) [inline src/modules/select/wallets/ledger.ts:211]: I think it should be removed

**Comment by jyap808** [comment]
Setting the HD Path for Ubiq also return an error:

image

**Comment by matextrem** [comment]
@jyap808 I don't think it is related with this PR since it is not working even in production https://reactdemo.blocknative.com/

**Comment by taylorjdawson** [comment]
@matextrem @nicosampler I am looking to get this merged in. Could you possibly provide me with the steps that you took when test this feature?

**Comment by matextrem** [comment]
Hey @taylorjdawson , thanks you so much for the reply!

In order to test it, I just connected a hardware wallet (I used Ledger nano S) with a dApp using this modified onboardjs version of the library.

This is the config I used: 

```
const onboard = Onboard({
   networkId: 1234
  //... other options
  walletSelect: {
     wallets:[
         {
             walletName: 'ledger',
             rpcUrl: RPC_URL,
            // We add a new custom parameter in order to specify the info needed for a custom network to work with HW 
            wallets.
             customNetwork: {
                  networkId: 246
                  genesis:  { 
                       hash: "0x0b6d3e680af2fc525392c720666cce58e3d8e6fe75ba4b48cb36bcc69039229b",        
                       timestamp: null,        
                       gasLimit: 6000000, 
                       difficulty: 131072, 
                       nonce: "0x0000000000000000", 
                       extraData: "",  
                       stateRoot: ""    
                  }
                  hardforks: [ 
                        { 
                           name: "chainstart", 
                           block: 0,        
                           consensus: "poa",        
                           finality: null        
                        },        
                        {        
                           name: "istanbul",        
                           block: 4922294,        
                           consensus: "poa",        
                           finality: null        
                        }    
                   ]
                  bootstrapNodes: [       
                        {        
                           ip: "35.181.30.169",        
                           port: 30303,        
                           id:
                    "cc67008e850c4b64702f1664f18704dd23980fb574138588904270739dca3a9c417ac3f97077b0f4e4872a5db109195f7137c156314dbcc724eabd809044553e",        
                          location: "",        
                          comment: ""        
                        },        
                        {        
                          ip: "35.181.104.81",        
                          port: 30303,        
                          id: "34f97743b9c07fac84c2db921337f006b5932e14c29df496951bdcfe3fce71e3d1504c25d66156dae01065623849b4fe940168c7e3f21c49b538af0cabb5805f",        
                          location: "",        
                          comment: ""        
                         },       
                         {       
                           ip: "3.1.233.174",        
                           port: 30303,        
                           id: "80e49825bb6bdfbf69576fd4a5f7bb173db98c08439bd7a019e2d42548afbb1f70232b74ede42dbd99cc0de3990a0101d49543c1b299027d2fd6f5a2c81eff75",        
                           location: "",        
                           comment: ""        
                          },        
                          {        
                            ip: "52.74.191.91",        
                            port: 30303,        
                            id: "1cd00f54ddaa41793d32500f0cb123d43e95f5ed18220c367d5a4db7d111e5fb59f250c397e40977562881875827f022d88a51e51171daaebf916beaa60e2139",        
                            location: "",        
                            comment: ""        
                           },        
                           {        
                             ip: "3.130.32.228",        
                             port: 30303,        
                             id: "3a8588bd883ea3bce606b08e7ca71d55717077cb3d25d3c00421ac371fbae286b94f58ba2e739b17b64c4d0d3e19f1b0ae2e47962703b42d7f36f1b9a448d4ea",        
                             location: "",        
                             comment: ""        
                             },      
                             {        
                               ip: "3.130.9.1",        
                               port: 30303,        
                               id: "69749dbe7e6f5fceaeb0a16f60353b3ec04825a12400e6d581fe980275e96fecb464276e61d79fc1628e448f1f74da2985d778ea54e0e8e7c64d980f2b3b6a94",        
                               location: "",        
                               comment: ""        
                              }    
                     ]
            }
         },
     ] 
  }
})
```
The configuration used above is the one related to `energy web chain` network, which is one of the network we need to support in Gnosis Safe. [Here](https://github.com/ethereumjs/ethereumjs-monorepo/pull/968) you can find the complete configuration json for EWC and Volta networks.

Let me know if you have any questions! 🚀

**Comment by taylorjdawson** [comment]
@matextrem Awesome! Could you outline steps needed to test on energy web chain network? Things I need are:
 - Gnosis url for the energy web chain network
 - Faucet or way to get some test volta

**Comment by matextrem** [comment]
Hey @taylorjdawson ! Let me send you what you asked me: 

- Gnosis url for the volta network (https://volta.gnosis-safe.io/app)
- Gnosis url for the energy web chain network (https://ewc.gnosis-safe.io/app)
- Faucet or way to get some test volta (https://voltafaucet.energyweb.org/)

**Review by taylorjdawson — APPROVED** [review]
I tested with Volta network and Ledger and it works well 👍🏾 
@aaronbarnardsound If you want to take a look and test before merging

**Review by aaronbarnardsound — APPROVED** [review]

**Pull request merged**